internal/client/client.go

```diff
@@ -175,126 +175,128 @@ func (c *CSAPI) SyncUntil(t *testing.T, since, key string, check func(gjson.Resu
-func (c *CSAPI) MustDoWithStatusRaw(t *testing.T, method string, paths []string, body []byte, contentType string, query url.Values, status int) *http.Response {
-	t.Helper()
-	res, err := c.DoWithAuthRaw(t, method, paths, body, contentType, query)
-	if err != nil {
-		t.Fatalf("CSAPI.MustDoWithStatusRaw %s %s error: %s", method, strings.Join(paths, "/"), err)
-	}
-	if res.StatusCode != status {
-		t.Fatalf("CSAPI.MustDoWithStatusRaw %s %s returned HTTP %d, expected %d", method, res.Request.URL.String(), res.StatusCode, status)
-	}
-	return res
-}
-
-// MustDo is the same as Do but fails the test if the response is not 2xx
+// MustDo will do the HTTP request and fail the test if the response is not 2xx
 func (c *CSAPI) MustDo(t *testing.T, method string, paths []string, jsonBody interface{}) *http.Response {
 	t.Helper()
-	res, err := c.DoWithAuth(t, method, paths, jsonBody)
-	if err != nil {
-		t.Fatalf("CSAPI.MustDo %s %s error: %s", method, strings.Join(paths, "/"), err)
-	}
+	res := c.DoFunc(t, method, paths, WithJSONBody(t, jsonBody))
 	if res.StatusCode < 200 || res.StatusCode >= 300 {
 		t.Fatalf("CSAPI.MustDo %s %s returned HTTP %d", method, res.Request.URL.String(), res.StatusCode)
 	}
 	return res
 }
 
-func (c *CSAPI) MustDoRaw(t *testing.T, method string, paths []string, body []byte, contentType string, query url.Values) *http.Response {
-	t.Helper()
-	res, err := c.DoWithAuthRaw(t, method, paths, body, contentType, query)
-	if err != nil {
-		t.Fatalf("CSAPI.MustDoRaw %s %s error: %s", method, strings.Join(paths, "/"), err)
+// WithRawBody sets the HTTP request body to `body`
+func WithRawBody(body []byte) RequestOpt {
+	return func(req *http.Request) {
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		// we need to manually set this because we don't set the body
+		// in http.NewRequest due to using functional options, and only in NewRequest
+		// does the stdlib set this for us.
+		req.ContentLength = int64(len(body))
 	}
-	if res.StatusCode < 200 || res.StatusCode >= 300 {
-		t.Fatalf("CSAPI.MustDoRaw %s %s returned HTTP %d", method, res.Request.URL.String(), res.StatusCode)
+}
+
+// WithContentType sets the HTTP request Content-Type header to `cType`
+func WithContentType(cType string) RequestOpt {
+	return func(req *http.Request) {
+		req.Header.Set("Content-Type", cType)
 	}
-	return res
 }
 
-// DoWithAuth a JSON request. The access token for this user will automatically be added.
-func (c *CSAPI) DoWithAuth(t *testing.T, method string, paths []string, jsonBody interface{}) (*http.Response, error) {
-	t.Helper()
-	var query url.Values
-	if c.AccessToken != "" {
-		query = url.Values{
-			"access_token": []string{c.AccessToken},
+// WithJSONBody sets the HTTP request body to the JSON serialised form of `obj`
+func WithJSONBody(t *testing.T, obj interface{}) RequestOpt {
+	return func(req *http.Request) {
+		t.Helper()
```

Comment:
Because only this function can fail the test if it cannot be JSON serialised. The others will never log or fail hence no need

Reply:
Oh! I thought the reason we were using `Helper` method was so that the calling function wouldn't appear in stack traces (potentially when other functions fail): https://golang.org/pkg/testing/#B.Helper. But it sounds like you're using it in a slightly different way?

Reply:
My point is that it cannot appear in a stack trace because it can never fail.

Reply:
It seems pointless to add `t` as an arg for `WithContentType` when it simply isn't used.

Reply:
Ah, I see what you're saying. I was concerned that something further down the stack would fail rather than this function, but that's not really likely to be the case. And the need to add the extra argument probably isn't worth it indeed.